### PR TITLE
Minimal Gossip tweaks

### PR DESCRIPTION
**Pull request by daviddrysdale**

### CODE REVIEW

**Comment by daviddrysdale** [comment]
Odd, Travis didn't seem to run. Closing and re-opening to see if that tickles it.

File: gossip/minimal/instance.go [hunk 1]
```diff
@@ -100,6 +100,26 @@ func NewGossiper(ctx context.Context, cfg *configpb.GossipConfig, hc *http.Clien
...
+			return nil, fmt.Errorf("failed to parse source log config: %v", err)
+		}
+		if _, ok := srcs[base.Name]; ok {
+			return nil, fmt.Errorf("duplicate source logs for name %s", base.Name)
+		}
+		srcs[base.Name] = &sourceLog{logConfig: *base}
+		knownSourceLogs.Set(1.0, base.Name)
+
+		// Assume that each source log has a new STH when polled.
+		sthRate := 1.0 // Default to 1 Hz if no minimum interval
```
Comment (RJPercival) [inline gossip/minimal/instance.go]: nit: Make this default a package const.

File: gossip/minimal/instance.go [hunk 2]
```diff
@@ -115,18 +135,14 @@ func NewGossiper(ctx context.Context, cfg *configpb.GossipConfig, hc *http.Clien
...
 		}
 		destPureHub.Set(isHub, hub.Name)
-	}
-	srcs := make(map[string]*sourceLog)
-	for _, lc := range cfg.SourceLog {
-		base, err := logConfigFromProto(lc, hc)
-		if err != nil {
-			return nil, fmt.Errorf("failed to parse source log config: %v", err)
+
+		submitRate := 1.0 // Default to 1 Hz if no minimum interval
```
Comment (RJPercival) [inline gossip/minimal/instance.go]: nit: Make this default a package const.

File: gossip/minimal/instance.go [hunk 3]
```diff
@@ -146,6 +162,9 @@ func logConfigFromProto(cfg *configpb.LogConfig, hc *http.Client) (*logConfig, e
 	if err != nil {
 		return nil, fmt.Errorf("failed to parse MinReqInterval: %v", err)
 	}
+	if interval <= 0 {
+		interval = time.Second
```
Comment (RJPercival) [inline gossip/minimal/instance.go]: nit: Make this default a package const.

File: gossip/minimal/instance.go [hunk 4]
```diff
@@ -170,6 +189,9 @@ func hubFromProto(cfg *configpb.HubConfig, hc *http.Client) (*destHub, error) {
...
```
Comment (RJPercival) [inline gossip/minimal/instance.go]: nit: Make this default a package const.

File: gossip/minimal/gossip.go [hunk 5]
```diff
@@ -274,6 +275,19 @@ type sthInfo struct {
...
+	waitChan := time.After(jitterWait)
+	select {
+	case <-ctx.Done():
+		glog.Infof("Retriever(%s): termination requested", src.Name)
+		return
+	case <-waitChan:
+		glog.V(1).Infof("Retriever(%s): wait for %v before starting...done", src.Name, jitterWait)
+	}
+
+	time.Sleep(jitterWait)
```
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply gossip/minimal/gossip.go]: Oops, no.  I had `time.Sleep()` then realized after the first Ctrl-C that I needed to make it interruptible, and ended up leaving in both versions -- thanks for spotting.

File: gossip/examples/deployment/docker/gosmin/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,14 @@
+FROM golang:1.10 as build
+
+ADD . /go/src/github.com/google/certificate-transparency-go
+WORKDIR /go/src/github.com/google/certificate-transparency-go
+
+RUN go get -v ./gossip/minimal/gosmin
```
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply gossip/examples/deployment/docker/gosmin/Dockerfile]: Done.

File: gossip/minimal/instance.go [hunk 7]
```diff
@@ -100,6 +100,26 @@ func NewGossiper(ctx context.Context, cfg *configpb.GossipConfig, hc *http.Clien
 		}
 	}
 
+	allSTHsRate := 0.0
+	srcs := make(map[string]*sourceLog)
+	for _, lc := range cfg.SourceLog {
+		base, err := logConfigFromProto(lc, hc)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse source log config: %v", err)
```
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply gossip/minimal/instance.go]: Done.

File: gossip/minimal/instance.go [hunk 8]
```diff
@@ -100,6 +100,26 @@ func NewGossiper(ctx context.Context, cfg *configpb.GossipConfig, hc *http.Clien
...
+		}
+		if _, ok := srcs[base.Name]; ok {
+			return nil, fmt.Errorf("duplicate source logs for name %s", base.Name)
```
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply gossip/minimal/instance.go]: Good idea, done.

File: gossip/minimal/instance.go (hunk 1 above)
Comment (daviddrysdale) [reply gossip/minimal/instance.go]: Done.

File: gossip/minimal/instance.go (hunk 2 above)
Comment (daviddrysdale) [reply gossip/minimal/instance.go]: Done.

File: gossip/minimal/instance.go (hunk 3 above)
Comment (daviddrysdale) [reply gossip/minimal/instance.go]: Done.

File: gossip/minimal/instance.go (hunk 4 above)
Comment (daviddrysdale) [reply gossip/minimal/instance.go]: Done.

**Pull request merged**